### PR TITLE
Ignore String constants by `RSpec/Describe`

### LINKED ISSUE #889: Suggestion: Disable `RSpec/DescribeClass` for versions
Hello!

I have (almost standard from `bundle gem`) gem version spec:

```ruby
# frozen_string_literal: true

describe 'FooBar::VERSION' do
  subject { FooBar::VERSION }

  it { is_expected.to match(/\d+\.\d+\.\d+/) }
end
```

And it causes an offense:

```
RSpec/DescribeClass: The first argument to describe should be the class or module being tested.	
```

Also, my DRY version:

```ruby
describe(const_name = 'FooBar::VERSION') do
  subject { Object.const_get(const_name) }
```

---

If I'd have `describe FooBar::VERSION do`, then output of specs (`--format=doc`) will contain `VERSION` value instead of constant name, and `described_class` will be `nil`.

### PR DESCRIPTION
**Pull request by AlexWayfer**
Resolve #889 

---

Before submitting the PR make sure the following are checked:

* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Updated documentation.
* [x] Added an entry to the [changelog](https://github.com/rubocop-hq/rubocop-rspec/blob/master/CHANGELOG.md) if the new code introduces user-observable changes.
* [x] The build (`bundle exec rake`) passes (be sure to run this locally, since it may produce updated documentation that you will need to commit).

### CODE REVIEW

**Comment by AlexWayfer** [comment]
I saw `def_node_matcher`, but I don't know how to insert Regexp and other logic into this, so I've created a regular private method.

File: spec/rubocop/cop/rspec/describe_class_spec.rb [hunk 1]
```diff
@@ -51,6 +51,21 @@
...
+    expect_no_offenses(<<-RUBY)
+      describe 'Some::VERSION' do
+      end
+    RUBY
+  end
+
+  it 'ignores assignment of non-class constants' do
+    expect_no_offenses(<<-RUBY)
+      describe(const_name = 'Some::VERSION') do
+        subject { Object.const_get(const_name) }
```
Comment (pirj) [inline spec/rubocop/cop/rspec/describe_class_spec.rb:64]: I've never seen this notation before you've mentioned it in the original ticket.
WDYT of:
```ruby
describe 'Some::VERSION' do
  subject { Object.const_get(self.class.description) }
```
We could get rid of the complex dance with `lvasgn`, and you'll get your case covered, and the default one for string description as well.

File: lib/rubocop/cop/rspec/describe_class.rb [hunk 2]
```diff
@@ -44,12 +44,22 @@ class DescribeClass < Cop
 
         def_node_matcher :shared_group?, SharedGroups::ALL.block_pattern
 
-        def on_top_level_describe(node, args)
+        def on_top_level_describe(node, (first_arg, _))
```
Comment (pirj) [inline lib/rubocop/cop/rspec/describe_class.rb:47]: What would be the best name for the first argument? Maybe `described_class`?

File: lib/rubocop/cop/rspec/describe_class.rb [hunk 3]
```diff
@@ -44,12 +44,22 @@ class DescribeClass < Cop
...
+          add_offense(first_arg)
+        end
+
+        private
+
+        def string_constant_describe?(first_arg)
+          value =
+            first_arg.lvasgn_type? ? first_arg.child_nodes.first : first_arg
+
+          value.str_type? && value.value =~ /(?:[A-Z]\w*(?:::)?)*[A-Z]\w*/
```
Comment (pirj) [inline lib/rubocop/cop/rspec/describe_class.rb:62]: This regexp is passing for `activeRecord` (unless it's too late to do code reviews).

File: spec/rubocop/cop/rspec/describe_class_spec.rb (hunk 1 above)
Comment (AlexWayfer) [reply spec/rubocop/cop/rspec/describe_class_spec.rb]: I think it'a better approach. Thank you!

File: lib/rubocop/cop/rspec/describe_class.rb (hunk 2 above)
Comment (AlexWayfer) [reply lib/rubocop/cop/rspec/describe_class.rb]: I've took this approach from `lib/rubocop/cop/rspec/describe_method.rb` (`second_arg`), but I think `described_value` is better, yes.

File: lib/rubocop/cop/rspec/describe_class.rb (hunk 3 above)
Comment (AlexWayfer) [reply lib/rubocop/cop/rspec/describe_class.rb]: Thank you! `^` and `$` have been added.

**Comment by AlexWayfer** [comment]
There are also other invalid cases with Regexp, wait a moment please.

**Comment by AlexWayfer** [comment]
> There are also other invalid cases with Regexp, wait a moment please.

Done. Tests have been added.

File: lib/rubocop/cop/rspec/describe_class.rb [hunk 4]
```diff
@@ -44,12 +44,20 @@ class DescribeClass < Cop
...
 
-          add_offense(args.first)
+          add_offense(described_value)
+        end
+
+        private
+
+        def string_constant_describe?(described_value)
+          described_value.str_type? &&
+            described_value.value =~ /^(?:(?:::)?[A-Z]\w*)+$/
```
Comment (pirj) [inline lib/rubocop/cop/rspec/describe_class.rb:60]: Deep research [here](https://stackoverflow.com/questions/53549477/regex-to-test-criteria-for-class-name), check it out.

This regex matched `'Active_Record::Base'`, I first thought it shouldn't, but it seems to be a valid class name though.

Is `?:` necessary? You can use `.match?` instead of `=~` to get a boolean.

File: spec/rubocop/cop/rspec/describe_class_spec.rb [hunk 5]
```diff
@@ -51,6 +51,33 @@
     RUBY
   end
 
+  context 'with non-class constants' do
```
Comment (pirj) [inline spec/rubocop/cop/rspec/describe_class_spec.rb:54]: WDYT of "with class or constant name as a string literal"?

File: spec/rubocop/cop/rspec/describe_class_spec.rb [hunk 6]
```diff
@@ -51,6 +51,33 @@
...
+    it 'ignores them' do
+      expect_no_offenses(<<-RUBY)
+        describe 'Some::VERSION' do
+          subject { Object.const_get(self.class.description) }
+        end
+      RUBY
+    end
+
+    it 'ignores root-constants with `::` at start' do
```
Comment (pirj) [inline spec/rubocop/cop/rspec/describe_class_spec.rb:63]: "ignores top-level constants"

File: spec/rubocop/cop/rspec/describe_class_spec.rb [hunk 7]
```diff
@@ -51,6 +51,33 @@
...
   end
 
+  context 'with non-class constants' do
+    it 'ignores them' do
+      expect_no_offenses(<<-RUBY)
+        describe 'Some::VERSION' do
+          subject { Object.const_get(self.class.description) }
+        end
+      RUBY
+    end
```
Comment (pirj) [inline spec/rubocop/cop/rspec/describe_class_spec.rb:61]: Maybe add a similar example with a regular class, e.g. `'Some::Thing'`? And also one without namespace `'Something'`.

File: lib/rubocop/cop/rspec/describe_class.rb (hunk 4 above)
Comment (AlexWayfer) [reply lib/rubocop/cop/rspec/describe_class.rb]: > This regex matched `'Active_Record::Base'`, I first thought it shouldn't, but it seems to be a valid class name though.

Yep, that's why I leave this.

---

> Is `?:` necessary?

No. I thought it's faster, but not:

```ruby
# frozen_string_literal: true

require 'bundler/setup'
Bundler.setup

require 'pry-byebug'

require 'benchmark'
require 'benchmark/ips'
require 'benchmark/memory'


puts '```ruby'
puts File.read(__FILE__).gsub("\t", '  ')
puts '```'
puts
puts '### Output'
puts
puts '```'


STRING = 'Active_Record::Base'


def with_non_captured_group
  STRING =~ /^(?:(?:::)?[A-Z]\w*)+$/
end

def with_captured_group
  STRING =~ /^(?:(?:::)?[A-Z]\w*)+$/
end


def test
  results = %i[
    with_non_captured_group
    with_captured_group
  ].each_with_object([]) do |method_name, array|
    array << (result = send method_name)
    puts "Result of #{method_name}:"
    pp result
    puts
  end
  uniq_size = results.uniq.size
  puts "Uniq results: #{uniq_size}"
  puts
  exit 1 if uniq_size > 1
end

test

Benchmark.ips do |x|
  x.report('with_non_captured_group') { with_non_captured_group }
  x.report('with_captured_group') { with_captured_group }

  x.compare!
end

Benchmark.memory do |x|
  x.report('with_non_captured_group') { 100.times { with_non_captured_group } }
  x.report('with_captured_group') { 100.times { with_captured_group } }

  x.compare!
end


puts '```'
```

### Output

```
Result of with_non_captured_group:
0

Result of with_captured_group:
0

Uniq results: 1

Warming up --------------------------------------
with_non_captured_group
                       185.982k i/100ms
 with_captured_group   183.201k i/100ms
Calculating -------------------------------------
with_non_captured_group
                          2.463M (± 3.0%) i/s -     12.461M in   5.064546s
 with_captured_group      2.430M (± 2.7%) i/s -     12.274M in   5.055925s

Comparison:
with_non_captured_group:  2462742.5 i/s
 with_captured_group:  2429607.0 i/s - same-ish: difference falls within error

Calculating -------------------------------------
with_non_captured_group
                        16.800k memsize (     0.000  retained)
                       100.000  objects (     0.000  retained)
                         0.000  strings (     0.000  retained)
 with_captured_group    16.800k memsize (     0.000  retained)
                       100.000  objects (     0.000  retained)
                         0.000  strings (     0.000  retained)

Comparison:
with_non_captured_group:      16800 allocated
 with_captured_group:      16800 allocated - same
```

Removed.

---



> You can use `.match?` instead of `=~` to get a boolean.

I wanted, but I saw `spec.required_ruby_version = '>= 2.3.0'`, and `String#match?` was introduced in Ruby 2.4.

File: spec/rubocop/cop/rspec/describe_class_spec.rb (hunk 5 above)
Comment (AlexWayfer) [reply spec/rubocop/cop/rspec/describe_class_spec.rb]: > WDYT of "with class or constant name as a string literal"?

OK, but "class or constant" is confusing (class is constant, is not it?), so I've left "with constant name as a String literal".

Comment (AlexWayfer) [reply spec/rubocop/cop/rspec/describe_class_spec.rb]: It turned into "when argument is a String literal".

File: spec/rubocop/cop/rspec/describe_class_spec.rb (hunk 6 above)
Comment (AlexWayfer) [reply spec/rubocop/cop/rspec/describe_class_spec.rb]: > "ignores top-level constants"

Replaced "root-constants" with "top-level constants", but they may be referenced without `::` at the start, so I left this.

File: spec/rubocop/cop/rspec/describe_class_spec.rb (hunk 7 above)
Comment (AlexWayfer) [reply spec/rubocop/cop/rspec/describe_class_spec.rb:85]: OK, added. Also another tests for empty strings and strings starting with numbers.

File: lib/rubocop/cop/rspec/describe_class.rb (hunk 4 above)
Comment (pirj) [reply lib/rubocop/cop/rspec/describe_class.rb]: Fair enough 👍

**Comment by pirj** [comment]
Good job @AlexWayfer! 👏 

@bquorning @Darhazer can you please glance over?

**Comment by AlexWayfer** [comment]
> Good job @AlexWayfer! :clap: 

Thank you. Your help is appreciated.

**Comment by AlexWayfer** [comment]
I have an idea: add a check like `Object.const_defined?`. What do you think? More strict.

**Comment by Darhazer** [comment]
> I have an idea: add a check like Object.const_defined?. What do you think? More strict.

It requires the app to be loaded. Rubocop does static analysis, so it doesn't know what constants are really defined. And if it needed to be defined before the specs are run, it would defeat the purpose of having it in a string

File: lib/rubocop/cop/rspec/describe_class.rb [hunk 8]
```diff
@@ -44,12 +44,20 @@ class DescribeClass < Cop
...
 
-          add_offense(args.first)
+          add_offense(described_value)
+        end
+
+        private
+
+        def string_constant_describe?(described_value)
+          described_value.str_type? &&
+            described_value.value =~ /^((::)?[A-Z]\w*)+$/
```
Comment (bquorning) [inline lib/rubocop/cop/rspec/describe_class.rb:60]: Could you please update the cop documentation with another “good” example, e.g. `'Some::Thing'`?

**Comment by AlexWayfer** [comment]
> It requires the app to be loaded. Rubocop does static analysis, so it doesn't know what constants are really defined. And if it needed to be defined before the specs are run, it would defeat the purpose of having it in a string

I didn't understand.

Expected behavior by this cop: use constants. Not stubbed constants. And test them. So, to use (and test) constant you have to load them, don't you?

File: lib/rubocop/cop/rspec/describe_class.rb (hunk 8 above)
Comment (AlexWayfer) [reply lib/rubocop/cop/rspec/describe_class.rb:65]: Thank you, I forgot about this, added.

Also with `subject { Object.const_get(self.class.description) }` — it can be helpful for others.

And I've added PR template checkbox about documentation in the same commit — I hope it's OK.

**Comment by Darhazer** [comment]
> Expected behavior by this cop: use constants. Not stubbed constants. And test them. So, to use (and test) constant you have to load them, don't you?

The point is, rubocop have no way of knowing about any constant defined outside of the file it inspects, it makes a static analysis that is limited to one file at a time, and doesn't actually run your application code.So it's not the right tool to verify that you are describing a real constant

**Comment by AlexWayfer** [comment]
> The point is, rubocop have no way of knowing about any constant defined outside of the file it inspects

And what? RuboCop lints `version_spec`… oh, I got it: although `version_spec` requires `lib/version`, but RuboCop doesn't execute this `require` while linting. OK, sorry.

**Comment by Darhazer** [comment]
> And what? RuboCop lints version_spec… oh, I got it: although version_spec requires lib/version, but RuboCop doesn't execute this require while linting. OK, sorry.

No problem. It's actually a limitation of Robocop actually.

**Comment by pirj** [comment]
Congratulations on your first contribution 🎉  @AlexWayfer

**Comment by AlexWayfer** [comment]
> Congratulations on your first contribution 🎉 @AlexWayfer

Thank you all, I'm glad to help (myself in the first place, yeah 😅 ).
